File: typescript/tracing/langchain/28_retriever_error.ts
import { BaseRetriever } from "@langchain/core/retrievers";

import { initRespan, shutdown, tracingConfig } from "./_shared";

class FailingRetriever extends BaseRetriever {
  lc_namespace = ["respan", "examples", "langchain"];

  async _getRelevantDocuments(query: string): Promise<never> {
    throw new Error(`retriever failed for ${query}`);
  }
}

export async function retrieverError(): Promise<void> {
  const runtime = await initRespan("typescript-langchain-retriever-error");
  const retriever = new FailingRetriever();

  try {
    await retriever.invoke("missing", tracingConfig(runtime, "retriever_error"));
  } catch (error) {
    console.log(`caught expected error: ${(error as Error).message}`);
  } finally {
    await shutdown(runtime);
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  await retrieverError();
}

File: python/tracing/langchain/28_retriever_error.py
"""Retriever error."""

from langchain_core.documents import Document
from langchain_core.retrievers import BaseRetriever

from _shared import init_telemetry, tracing_config


class FailingRetriever(BaseRetriever):
    def _get_relevant_documents(self, query: str, *, run_manager=None) -> list[Document]:
        raise RuntimeError(f"retriever failed for {query}")


def retriever_error() -> None:
    telemetry = init_telemetry("langchain-retriever-error")
    retriever = FailingRetriever()
    try:
        retriever.invoke("missing", config=tracing_config("retriever_error"))
    except RuntimeError as exc:
        print(f"caught: {exc}")
if __name__ == "__main__":
    retriever_error()
